core/types/block_ext_test.go

```diff
@@ -106,3 +111,410 @@ func assertDifferentPointers[T any](t *testing.T, a *T, b any) {
+	assert.Equal(t, wantBodyExtra, bodyExtra)
+
+	exportedFieldsPointToDifferentMemory(t, blockExtras, bodyExtra)
+}
+
+func TestBlockGetters(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name               string
+		headerExtra        *HeaderExtra
+		blockExtra         *BlockBodyExtra
+		wantExtDataGasUsed *big.Int
+		wantBlockGasCost   *big.Int
+		wantVersion        uint32
+		wantExtData        []byte
+	}{
+		{
+			name:        "empty",
+			headerExtra: &HeaderExtra{},
+			blockExtra:  &BlockBodyExtra{},
+		},
+		{
+			name: "fields_set",
+			headerExtra: &HeaderExtra{
+				ExtDataGasUsed: big.NewInt(1),
+				BlockGasCost:   big.NewInt(2),
+			},
+			blockExtra: &BlockBodyExtra{
+				Version: 3,
+				ExtData: ptrTo([]byte{4}),
+			},
+			wantExtDataGasUsed: big.NewInt(1),
+			wantBlockGasCost:   big.NewInt(2),
+			wantVersion:        3,
+			wantExtData:        []byte{4},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			header := &Header{}
+			SetHeaderExtra(header, test.headerExtra)
+
+			block := NewBlock(header, nil, nil, nil, stubHasher{})
+			extras.Block.Set(block, test.blockExtra)
+
+			extData := BlockExtData(block)
+			assert.Equal(t, test.wantExtData, extData)
+
+			version := BlockVersion(block)
+			assert.Equal(t, test.wantVersion, version)
+
+			extDataGasUsed := BlockExtDataGasUsed(block)
+			assert.Equal(t, test.wantExtDataGasUsed, extDataGasUsed)
+
+			blockGasCost := BlockGasCost(block)
+			assert.Equal(t, test.wantBlockGasCost, blockGasCost)
+		})
+	}
+}
+
+func TestNewBlockWithExtData(t *testing.T) {
+	t.Parallel()
+
+	// This transaction is generated beforehand because of its unexported time field being set
+	// on creation.
+	testTx := NewTransaction(0, common.Address{1}, big.NewInt(2), 3, big.NewInt(4), []byte{5})
+
+	tests := []struct {
+		name      string
+		header    func() *Header
+		txs       []*Transaction
+		uncles    func() []*Header
+		receipts  []*Receipt
+		extdata   []byte
+		recalc    bool
+		wantBlock func() *Block
+	}{
+		{
+			name: "empty",
+			header: func() *Header {
+				header := &Header{}
+				SetHeaderExtra(header, &HeaderExtra{})
+				return header
+			},
+			wantBlock: func() *Block {
+				header := &Header{}
+				SetHeaderExtra(header, &HeaderExtra{})
+				block := NewBlock(header, nil, nil, nil, stubHasher{})
+				blockExtra := &BlockBodyExtra{ExtData: ptrTo([]byte{})}
+				extras.Block.Set(block, blockExtra)
+				return block
+			},
+		},
+		{
+			name: "with_recalc",
+			header: func() *Header {
+				header := &Header{}
+				extra := &HeaderExtra{
+					ExtDataHash: common.Hash{1}, // should be overwritten
+				}
+				SetHeaderExtra(header, extra)
+				return header
+			},
+			extdata: []byte{2},
+			recalc:  true,
+			wantBlock: func() *Block {
+				header := &Header{}
+				extra := &HeaderExtra{
+					ExtDataHash: common.HexToHash("0xf2ee15ea639b73fa3db9b34a245bdfa015c260c598b211bf05a1ecc4b3e3b4f2"),
```

Comment:
```suggestion
					ExtDataHash: CalcExtDataHash([]byte{2}),
```

Otherwise it's just a magic number.

Reply:
I think this gets computed internally with RLP, so you cannot set what you want really.